### PR TITLE
improve millis IRQ with asm

### PR DESCRIPTION
**Pull request by MX682X**
Quick improvement for millis IRQ. Improves cycles and flash by about 5-15%

for comparison, that was the generated asm before (I wish I knew how to tell gcc that ram is more costly than register manipulation)

```
 8aa:	1f 92       	push	r1
 8ac:	0f 92       	push	r0
 8ae:	0f b6       	in	r0, 0x3f	; 63
 8b0:	0f 92       	push	r0
 8b2:	11 24       	eor	r1, r1
 8b4:	8f 93       	push	r24
 8b6:	9f 93       	push	r25
 8b8:	af 93       	push	r26
 8ba:	bf 93       	push	r27
 8bc:	80 91 a7 40 	lds	r24, 0x40A7	; 0x8040a7 <timer_millis>
 8c0:	90 91 a8 40 	lds	r25, 0x40A8	; 0x8040a8 <timer_millis+0x1>
 8c4:	a0 91 a9 40 	lds	r26, 0x40A9	; 0x8040a9 <timer_millis+0x2>
 8c8:	b0 91 aa 40 	lds	r27, 0x40AA	; 0x8040aa <timer_millis+0x3>
 8cc:	01 96       	adiw	r24, 0x01	; 1
 8ce:	a1 1d       	adc	r26, r1
 8d0:	b1 1d       	adc	r27, r1
 8d2:	80 93 a7 40 	sts	0x40A7, r24	; 0x8040a7 <timer_millis>
 8d6:	90 93 a8 40 	sts	0x40A8, r25	; 0x8040a8 <timer_millis+0x1>
 8da:	a0 93 a9 40 	sts	0x40A9, r26	; 0x8040a9 <timer_millis+0x2>
 8de:	b0 93 aa 40 	sts	0x40AA, r27	; 0x8040aa <timer_millis+0x3>
 8e2:	81 e0       	ldi	r24, 0x01	; 1
 8e4:	80 93 26 0b 	sts	0x0B26, r24	; 0x800b26 <__TEXT_REGION_LENGTH__+0x7e0b26>
 8e8:	bf 91       	pop	r27
 8ea:	af 91       	pop	r26
 8ec:	9f 91       	pop	r25
 8ee:	8f 91       	pop	r24
 8f0:	0f 90       	pop	r0
 8f2:	0f be       	out	0x3f, r0	; 63
 8f4:	0f 90       	pop	r0
 8f6:	1f 90       	pop	r1
 8f8:	18 95       	reti
```

### CODE REVIEW

**Comment by SpenceKonde** [comment]
That trick for economizing register use in the ISR is mighty clever.
Your point that this can be done better in assembly is well taken, however I'm not sure I belive it works as written. you clobber r24, but don't list it as a clobber - I don't know if the compiler is smart enough to realize it needs to save and restore r24, meaning that whatever was in r24 in the the code that was interrupted would be corrupted, and I don't think it works at 2 MHz, I need to doublecheck how often the millis interrupt is supposed to fire at 2 MHz, I thought it was still once per 2ms. But most concerned about what's going on with the SBCI's - Does that work correctly? It can't possibly work right as written at 1 MHz, right? Need to do that out with pencil and paper to see what exactly it will do. 

In any event, with no use of r0 or r1 and no C, just inline asm, the TCB ISRs should be cordoned off so that they can get naked, prologue would push r24 to save it, load sreg to r24 with in, push again, push r26 and r27, then proceed with the considerably more efficient math (with I suspect some corrections) and then the epilogue would just be pop r27 pop r26 pop r24 out r24, sreg, pop r24 to restore it and reti.  That would save 6 more clocks if r24 is handled correctly - which as I said, it doesn't appear to be (in which case, millis isr would shit on r24 every millisecond and as that's one of the most used registers, all but the simplest code would malfunction 1000 times per second).

**Comment by MX682X** [comment]
>  you clobber r24, but don't list it as a clobber 

since the ISR also resets the ISR flag of the timer, it will allocate r24 automatically, but have added the clobber declaration anyway.

> and I don't think it works at 2 MHz,

There is no board option for 2MHz

> In any event, with no use of r0 or r1 and no C

Well, there's the problem, there is still C. I haven't touched the rest of the function (TCA/TCD/RTC) + Timer ISR flag reset.

**Comment by SpenceKonde** [comment]
Okay, I've fixed the 2MHz case (it does do two milliseconds)
I also made the ISR naked, because at this point we really might as well. 

Previously, counting time taken to enter the ISR to the time it resumed interrupted function was 60-61 clocks, and it used 40 words of flash
Now, it is 44-45 clocks by the same measure, a savings of 16 clocks, and 28 words of flash, a savings of 12. 
Changing the math saved 8 clocks doing the math and 2 in the epilog but cost one before the math started for 9 clocks. and 7 words of flash in the math and you save 1 in epi but lose 1 before doing the math, for 7 words of flash. My changes saved an additional 7 clocks and 5 words. 4 and 2 in the epilog and 3 and 3 in the prologue. 

Mine was certainly not worth the trouble alone, and it had a greater cost in ugliness for smaller returns, a 25% executiontime savings and 30% flash savings from the combined effort somehow seems more worthwhile

So now the TCBs get something that looks something like this.  
```
      "push       r26"
      "push       r27"
      "ldi           r26, 0x??" <---- these are a little sketchy to do naked, but is known to work from serial, assuming you 
      "ldi           r27, 0x??"           clear the regs first, and all you ask the compiler to do is to generate LDI and/or LDS isns. 
      "push       r24"
      "in         r24,   0x3F"
      "push       r24"
      "ld         r24,      X"
      "subi       r24, %[DEC]"
      "st          X+,    r24"
      "ld         r24,      X"
      "sbci       r24,   0xFF"
      "st          X+,    r24"
      "ld         r24,      X"
      "sbci       r24,   0xFF"
      "st          X+,    r24"
      "ld         r24,      X"
      "sbci       r24,   0xFF"
      "st           X,    r24"
      "ldi        r24,   0x01"
      "sts        0x????, r24" <--- the way I get this is ugly, but unquestionably legal.
      "pop        r24"
      "out       0x3F,    r24"
      "pop        r24"
      "pop        r27"
      "pop        r26"
      "reti"                        
```
Your change alone would have generated something like this; my change just eliminates the pointless r0/r1 manipulations.
```
 8aa: 	1f 92       	push	r1 
 8ac: 	0f 92       	push	r0
 8ae:	0f b6       	in     	r0, 0x3f	; 63
 8b0:	0f 92       	push	r0
 8b2:	11 24      	eor   	r1, r1
 8b4:	8f 93       	push	r24
 8b8:	af 93       	push	r26
 8ba:	bf 93       	push	r27
 8bc:	xx xx       	ldi    	r26, 0x??	; ?
 8be:	xx xx       	ldi    	r27, 0x??	; ?
 8c0: 	xx xx      	ld     	r24, X
 8c2: 	xx xx      	subi    	r24, 0xFF	; 255
 8c4: 	xx xx      	st     	r24, X+
 8c6: 	xx xx      	ld     	r24, X
 8c8: 	xx xx      	sbci    	r24, 0xFF	; 255
 8ca: 	xx xx      	st     	r24, X+
 8cc: 	xx xx      	ld     	r24, X
 8ce: 	xx xx      	sbci    	r24, 0xFF	; 255
 8d0: 	xx xx      	st     	r24, X+
 8d2: 	xx xx      	ld     	r24, X
 8d4:	xx xx      	sbci    	r24, 0xFF	; 255
 8d6:	xx xx      	st     	r24, X+
 8d8:	xx xx       	ldi    	r24, 0x01	; 1
 8da:	80 93 26 0b 	sts	0x0B26, r24	; 0x800b26 <__TEXT_REGION_LENGTH__+0x7e0b26> 
 8de:	bf 91       	pop  	r27
 8e0:	af 91       	pop  	r26
 8e2:	8f 91       	pop  	r24
 8e4: 	0f 90       	pop  	r0
 8e6: 	0f be       	out   	0x3f, r0	; 63
 8e8: 	0f 90       	pop  	r0
 8ea: 	1f 90       	pop  	r1
 8ec: 	18 95      	reti
```

Also
```
sts	0x0B26, r24
```
That aint from megaTinyCore, that's DxCore with millis on TCB2 :-P

This change will of course be going in over there as well. 
While I was there I also fixed some of the missing indentation that made the structure of the conditionals  hard to follow (a situation hardly improved by the massive comment blocks in micros inside the same conditional complex.

**Comment by MX682X** [comment]
Just a thought, but is it worth to keep the SREG in say r0 while we are in the ISR? Instead of pushing and popping it too
